### PR TITLE
CustomReader

### LINKED ISSUE #10873: Custom Reader
In our design, `reader` is used to read data from files as well as doing data preprocessing. A few basic and common data preprocessing has been implemented by us(batching, shuffling,...) in the form of `DecoratedReader`. However, there are unlimited ways to do data preprocessing, so it is necessary to introduce a 'custom reader', with which users are able to config their own preprocessing pipelines. 

`CustomReader` is also a `DecoartedReader`. Unlike other decorated readers, a `CustomReader` holds its own sub-block. Users can config their own preprocessing pipelines in the sub-block, and the sub-block will be executed in the `CustomReader`'s `ReadNext()` function.

### PR DESCRIPTION
**Pull request by JiayiFeng**
fixes #10873

### CODE REVIEW

File: paddle/fluid/operators/reader/create_custom_reader_op.cc [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  out->clear();
+  std::vector<framework::LoDTensor> underlying_outs;
+  reader_->ReadNext(&underlying_outs);
+  if (underlying_outs.empty()) {
+    // There is not next data.
+    return;
+  }
+  PADDLE_ENFORCE(
+      source_var_names_.size() == underlying_outs.size() &&
+          sink_var_names_.size() == underlying_outs.size(),
```
Comment (reyoung) [inline paddle/fluid/operators/reader/create_custom_reader_op.cc]: sink_var_names maybe not equal with source_var_names.size()

File: paddle/fluid/operators/reader/create_custom_reader_op.cc [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+      "The size of source_var_names(%d), the size of sink_var_names(%d) and "
+      "the size of underlying_outs(%d) are not consistent. Each feeding "
+      "element must have its own source and sink variable.",
+      source_var_names_.size(), sink_var_names_.size(), underlying_outs.size());
+  // The scope for CustomReader's sub-block should be independent and shouldn't
+  // be any other computation scope's child. Otherwise, data preprocessing and
+  // compution cannot be concurrent.
+  auto* scope = new framework::Scope();
```
Comment (reyoung) [inline paddle/fluid/operators/reader/create_custom_reader_op.cc]: Well, it is no need to `new Scope()`. Just 

```C++
Scope scope;
```
is cool

File: paddle/fluid/operators/reader/create_custom_reader_op.cc (hunk 1 above)
Comment (JiayiFeng) [reply paddle/fluid/operators/reader/create_custom_reader_op.cc]: Fixed. Thanks!

File: paddle/fluid/operators/reader/create_custom_reader_op.cc (hunk 2 above)
Comment (JiayiFeng) [reply paddle/fluid/operators/reader/create_custom_reader_op.cc]: Good idea.

File: python/paddle/fluid/layers/io.py [hunk 3]
```diff
@@ -514,3 +514,81 @@ def read_file(file_obj):
...
+
+        source_shapes = self.underlying_reader.desc.shapes()
+        source_dtypes = self.underlying_reader.desc.dtypes()
+        source_lod_levels = self.underlying_reader.desc.lod_levels()
+        self.source_var_names = []
+        source_vars = []
+        for idx in xrange(len(source_shapes)):
+            self.source_var_names.append(unique_name("preprocessor_source"))
+            source_vars.append(self.main_prog.current_block().create_var(
+                name=self.source_var_names[-1],
```
Comment (reyoung) [inline python/paddle/fluid/layers/io.py]: Should here be `self.source_var_names[-1]` or `self.source_var_names[idx]`.

File: python/paddle/fluid/layers/io.py [hunk 4]
```diff
@@ -514,3 +514,81 @@ def read_file(file_obj):
...
+            raise RuntimeError(
+                "Preprocessor.inputs() can only be invoked inside the sub-block."
+            )
+
+        source_shapes = self.underlying_reader.desc.shapes()
+        source_dtypes = self.underlying_reader.desc.dtypes()
+        source_lod_levels = self.underlying_reader.desc.lod_levels()
+        self.source_var_names = []
+        source_vars = []
+        for idx in xrange(len(source_shapes)):
```
Comment (reyoung) [inline python/paddle/fluid/layers/io.py]: Maybe more pythonic way
```python
source_var_names = (unique_name("preprocessor_source") for _ in xrange(len(source_shapes)))
for var_name, shape, dtype in zip(source_var_names, source_shapes, source_dtypes):
	pass
```

File: paddle/fluid/operators/reader/create_custom_reader_op.cc [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    tensor->ShareDataWith(underlying_outs[i]);
+    tensor->set_lod(underlying_outs[i].lod());
+  }
+  // 2. Run the sub-block.
+  exe_.Run(program_, scope, sub_block_id_, false, true);
+  // 3. Copy LoDTensors from sink variables to out.
+  out->resize(sink_var_names_.size());
+  for (size_t i = 0; i < sink_var_names_.size(); ++i) {
+    framework::Variable* var = scope->FindVar(sink_var_names_[i]);
+    PADDLE_ENFORCE_NOT_NULL(var);
```
[the remark this answers is not in the document]
Comment (JiayiFeng) [reply paddle/fluid/operators/reader/create_custom_reader_op.cc]: Cool!

File: python/paddle/fluid/layers/io.py (hunk 4 above)
Comment (JiayiFeng) [reply python/paddle/fluid/layers/io.py]: Amazing!

**Pull request merged**